src/phoenix/server/api/schema.py

```diff
@@ -203,44 +204,39 @@ def hdbscan_clustering(
     def spans(
         self,
         info: Info[Context, None],
+        time_range: Optional[TimeRange] = UNSET,
         trace_ids: Optional[List[ID]] = UNSET,
         first: Optional[int] = 50,
         last: Optional[int] = UNSET,
         after: Optional[Cursor] = UNSET,
         before: Optional[Cursor] = UNSET,
         sort: Optional[SpanSort] = UNSET,
         root_spans_only: Optional[bool] = False,
+        filter_condition: Optional[str] = None,
     ) -> Connection[Span]:
-        if info.context.traces is None:
-            spans = []
-        else:
-            df = info.context.traces._dataframe
-            if trace_ids:
-                df = df.loc[df.loc[:, TRACE_ID].isin(trace_ids)]
-            if root_spans_only:
-                df = df.loc[df.loc[:, PARENT_ID].isna()]
-            sort = (
-                SpanSort(col=SpanColumn.startTime, dir=SortDir.asc)
-                if not sort or sort.col.value not in df.columns
-                else sort
-            )
-            # Convert dataframe rows to Span objects
+        spans: Iterable[s.Span] = ()
+        if (traces := info.context.traces) is not None:
+            try:
+                predicate = Filter(filter_condition) if filter_condition else None
+            except SyntaxError as e:
+                raise Exception(f"invalid filter condition: {e.msg}") from e  # TODO: add details
             spans = (
-                []
-                if df.empty
-                # When df is empty, `pandas.DataFrame.apply` sometimes returns
-                # DataFrame when it should return Series. It's unclear why that
-                # happens.
-                else sort.apply(df)  # type: ignore
-                .apply(
-                    to_gql_span,
-                    axis=1,
+                chain.from_iterable(
+                    map(traces.get_trace, map(UUID, trace_ids)),
+                )
+                if trace_ids
+                else traces.get_spans(
+                    start_time=time_range.start if time_range else None,
+                    stop_time=time_range.end if time_range else None,
+                    root_spans_only=root_spans_only,
                 )
```

Comment:
the level of conditionals and nesting here makes this kinda hard to read - I'd consider just using simpler syntax to make it a bit readable.